### PR TITLE
[WIP][CIAPP] CI Visibility Mode v1

### PR DESCRIPTION
**Pull request by drodriguezhdez**
This PR is a refactor to the CI App implementation into a new CI Visibility Mode.

This mode includes:

* Custom initialization entrypoint. (TBD)
* Testing framework integrations are now only enabled when running in CI Visibility mode (TBD)

### CODE REVIEW

File: dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/writer/CIWriter.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package datadog.trace.ci.writer;
```
Comment (dougqh) [inline dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/writer/CIWriter.java:1]: I don't like the use of inheritance.  I think this is better handled through configuration & composition.

File: dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/writer/CIAgentWriter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import datadog.trace.common.writer.DDAgentWriter;
+import datadog.trace.core.DDSpan;
+import java.util.List;
+
+/**
+ * This writer is specific for CI Visibility using Datadog Agent. It is based on the {@code
+ * DDAgentWriter} class applying the following logic: - Only traces with span.type=test (local root
+ * span) are sent. - All children spans contains the origin tag.
+ */
+public class CIAgentWriter implements CIWriter {
```
Comment (dougqh) [inline dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/writer/CIAgentWriter.java:12]: I don't like the use of inheritance. I think this is better handled through configuration & composition.

File: dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/metrics/CIStatsDClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+public class CIStatsDClientFactory {
+
+  private static final String LANG_STATSD_TAG = "lang";
+  private static final String LANG_VERSION_STATSD_TAG = "lang_version";
+  private static final String LANG_INTERPRETER_STATSD_TAG = "lang_interpreter";
+  private static final String LANG_INTERPRETER_VENDOR_STATSD_TAG = "lang_interpreter_vendor";
+  private static final String TRACER_VERSION_STATSD_TAG = "tracer_version";
+
+  public static StatsDClient createCiStatsDClient(final Config config) {
+    if (!config.isHealthMetricsEnabled()) {
```
Comment (dougqh) [inline dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/metrics/CIStatsDClientFactory.java:22]: It looks like we're repeating code from elsewhere in the tracer.  Why?

File: dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/CITracerFactory.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import datadog.trace.core.DDSpan;
+
+/**
+ * Tracer Factory for CI Visibility It creates the tracer with the correct configuration to be used
+ * in the CI Visibility mode.
+ */
+public class CITracerFactory {
+
+  public static CITracer createCITracer(
+      final Config config, SharedCommunicationObjects commObjects) {
```
Comment (dougqh) [inline dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/CITracerFactory.java:22]: I don't think many parallel objects are necessary.  Just configuration and a different composition of supporting objects for the Tracer.

File: dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/CITracer.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import datadog.trace.context.ScopeListener;
+import datadog.trace.context.TraceScope;
+import datadog.trace.core.CoreTracer;
+
+/**
+ * This tracer is specific for CI Visibility The CITracer is backed by an instance of the {@code
+ * CoreTracer} class configured to be used in the CI Visibility mode.
+ */
+public class CITracer implements AgentTracer.TracerAPI {
+
```
Comment (dougqh) [inline dd-trace-core/ci-visibility/src/main/java/datadog/trace/ci/CITracer.java:21]: I don't want completely different Tracer for CI.

**Review by dougqh — CHANGES_REQUESTED** [review]
I think this change set is headed in the wrong direction.
I don't want to create a whole set of parallel classes for CI.

To the best of my knowledge, CI only needs a slightly differently configured sending pipeline.  That can be accomplished through configuration objects and strategy composition.

**Comment by drodriguezhdez** [comment]
https://github.com/DataDog/dd-trace-java/pull/3196

**Pull request closed without merging**